Blas mpt proof

blasrodri opened the pull request.

blasrodri committed:
create proof and verification

first steps :)

blasrodri committed:
few changes

blasrodri committed:
expose proof in rpc

akhi3030 commented on core/store/src/trie/mod.rs:
```diff
@@ -757,6 +766,97 @@ impl Trie {
+        &self,
+        root: &CryptoHash,
+        key: &[u8],
+    ) -> Result<Option<Vec<Level>>, StorageError> {
```
slightly off-topic.  Wouldn't `Result<Vec<Level>, StorageError>` suffice here?  Isn't a vector of length 0 sufficient to capture the `None` situation or do we need to represent them differently?

blasrodri replied:
We can default to that.

akhi3030 commented on runtime/runtime/src/state_viewer/mod.rs:
```diff
@@ -157,7 +157,20 @@ impl TrieViewer {
+
+        let proof = trie
+            .get_proof(&root, &query)?
+            .unwrap_or_else(|| vec![])
```
@blasrodri : moving conversation from zulip here.  Do I understand correctly that you are getting a `None` here?

akhi3030 replied:
In which case, we need to study `get_proof()` above to see under which conditions it can return `Ok(None)`.  Maybe try some `printf` debugging to see which case we are hitting and try to reason about why we are hitting that case.

blasrodri replied:
I think so. But I need to understand what is the expected result. 

```rs
    let result = trie_viewer.view_state(&state_update, &alice_account(), b"").unwrap();
    assert_eq!(result.proof, Vec::<String>::new());
```

when passing empty prefix, the proof is empty. Same with prefix being "xyz". Only when it is "test123" it gets the proof. I wonder whether this is correct, or should be modified. @akhi3030

akhi3030 replied:
@blasrodri: I am stumped too!  I don't see any obvious issues with the implementation.  We probably need some other folks from NEAR to weigh in on this.  I suggest that you post in #general channel on zulip.

mzhangmzz committed:
Make apply_chunks async (#6968)

This PR makes block processing asynchronous by moving apply_chunks to another thread. This changes the interface of block processing from `process_block`, `process_block_single` to `start_process_block_async` and `postprocess_ready_blocks`. 

A lot of changes in this PR is because of the callbacks on accepted blocks. When a block is accepted (finished processing), Chain needs to call `check_orphans` to see if this block unlock any orphans in the orphan pool so that these blocks are ready to be processed. Client also needs to call `on_block_accepted` on every accepted block. After the change, `start_process_block_async` won't trigger any of these callbacks. These are moved to `postprocess_ready_blocks` because that is the only place where blocks processing finishes. 

Test Plan:
- Unit test using Client. I modified test_processing_chunks_sanity and added a new test test_processing_blocks_async
- Some tests using ClientActor, probably a python test (TODO)
- Testing using localnet (TODO)

matklad committed:
refactor: simplify PROTOCOL_VERSION declaration (#7185)

don't need a macro here

jakmeier committed:
feat: estimate fn call action without loading cost (#6996)

`action_function_call_*` costs are currently used to also cover contract
loading. But we want to give `wasm_contract_loading_*` costs meaningful
values and consequently reduce `action_function_call_*`.

This adjusts the `ActionFunctionCallBase` estimation to not include
contract loading costs. The difference is rather singificant since the
test contract used before was almost 100kB in size.

This change also introduces an approach of subtracting receipt overhead
completely, by executing multiple actions into one receipt.

firatNEAR committed:
Fix/rosetta rpc account balance (#7186)

jakmeier committed:
feat: stabilize precharged contract loading gas (#7177)

# Feature to stabilize
The gas cost for loading a contract executable used to be charged afterwards. However, the general rule for gas charges is that they should be charged before the cost occurs. This is only noticeable when the execution fails during that step. But it is nevertheless important, as an attacker could potentially abuse this to fill up block space without being charged for it.

Note that "loading" in this context is a technical term for preparing and populating all memory regions for an executable. Reading the data from the database is NOT included in this.

On a side note, the actual gas parameter values have not been updated, yet. We know that `wasm_contract_loading_bytes` in particular is too low. But it is counter balanced by `action_function_call` which is higher than it needs to be. Stabilizing this feature paves the floor to adjust those parameters.

Main implementation PR: #6772
Main issue: #5962

# Testing and QA
## Tests
The original PR (#6772) includes tests on the VM level that ensure old versions are unaffected while new versions charge the cost correctly.
Integration tests in #7169 checks the same again but on the apply-block level.

## Impact on existing smart contracts
This change does not affect anything except for function calls that fail during preparation. Since those were already failing, it should not impact anyone if the error type and amount of gas charged changes.

# Checklist
- [x] Link to nightly nayduck: https://nayduck.near.org/#/run/2562
- [x] Update CHANGELOG.md to include this protocol feature in the `Unreleased` section.

ChaoticTempest committed:
chore: Enable M1 builds on master (#7173)

This will enable the M1 GH pipeline to build and publish M1 binaries of nearcore. Thanks @vlad-kopcil for putting this together

matklad committed:
test: unignore wasmer2 upgrade test (#7189)

It was temporarilty ignored to ship async block processing.

The reason why this test fails after the change is because, in the test,
we inspect the logs, and, now that the work happens in a different
threads, tracing loses its thread-local context and we fail to capture
the span.

In the ideal world, tracing&rayon would just magically work this out
between themselves, but today it seems that the best option is for us to
help them.

See also discussion at

https://discord.com/channels/500028886025895936/627649734592561152

matklad committed:
test: simplify code a bit (#7190)

matklad committed:
test: tiny simplification in setup_mock_all_validators (#7191)

It probably takes the crown as the scariest function in our codebase,
and it'd be great to simplify this significantly. Let's start somewhere.

matklad committed:
tests: cleanup to use HashMap "literal" (#7193)

matklad committed:
test: simplify setup_mock_validators a bit (#7192)

Ekleog committed:
fuzz: copy all cargo-fuzz targets to cargo-bolero integration tests (#7143)

This gets us ready for nayduck changes that will run cargo-bolero instead of cargo-fuzz. Once the nayduck changes are complete we'll be able to remove the cargo-fuzz targets

Ekleog committed:
Disable the secondary stack limit for fuzzing (#7195)

mina86 committed:
store: remove unused map_col_from_key function (#7158)

The function is, I assume, present in the code in case it’s going to
be useful in future migrations.  However, there already is map_col
function which has very similar behaviour so in the future if someone
needs feature provided by map_col_from_key they won’t have issues
adapting map_col for their purpose.

While at it, do some minor cleanups in migration module.

blasrodri committed:
update name of Level to TrieMerkleProofItem

blasrodri committed:
Merge branch 'master' into blas-mpt-proof

mina86 committed:
Remove duplicated serialisation formats implementations (#7201)

Fixes: https://github.com/near/nearcore/issues/3384

matklad committed:
test: simplify mocking in setup_mock_all_validators (#7194)

There are many lovable aspects to `setup_mock_all_validators`! What this PR deals with is "tying the knot" infra. `setup_mock_validator` is based on creating `n` validators and setting up the mock network between them. To create a mock network, you need all `n` validators, and to create a validator, you need a mock network. The cycle is broken at runtime, in three different ways: 

* inside `setup_mock_all_validators`, we have `locked_connectors` 
* at some call-sites, we overwrite the whole `peer_manager_mock`
* yet at other call-sites, we essentially create a copy of `locked_connectors`

This PR unifies that to a single infra: 

* there's a single `Arc<OnceCell<Vec<Addr<Client>>>>` inside `setup_mock_all_validators`
* `OnceCell::wait` is used as a barrier 
* The `&[Addr<CLient>]` are passed in as an argument to `peer_manager_mock`, so that it doesn't have to re-invent a way to get a hold of those

This changes the signature of the function, so the diff is pretty large. It is designed to be reviewed per-commit.

ChaoticTempest committed:
Fix m1 build issue (#7199)

pompon0 committed:
renamed Validator to AccountData (#7206)

https://nearinc.atlassian.net/browse/CPR-94

jakmeier committed:
feat: new estimator variance check (#7197)

The condition to check for high variance of measurements so far has been
that no single sample is further than 15% from the mean away. We often
missed that when running with 5 iterations and one warm-up iteration.
The counter-intuitive side of this check is that is becomes less likely
that variance is considered low enough, if more iterations are added.

The new condition computes actual variance and standard deviation.
Now we expect that standard deviation is lower than 10%.

Also add an outer loop to block overhead computation. Usually there
is a loop + a number of actions inside a block. But this estimation
has empty blocks and therefore each measurement so far was
just a single apply block call. Grouping 10 empty blocks together
for each measurement and adding an outer loop for `--iters` reduces 
the overall variance.

matklad committed:
test: more useful assert (#7211)

Got this one spuriously(?) firing in some test with nightly features.
Let's make the failure a bit more helpful here!

mina86 committed:
trie_key: don’t store single bytes as slices (#7200)

trie_key module stores separators and keys as slices even though they
are all just one-byte long.  There are parts of the code which assume
the length to be one and won’t work if the length ever changes.

Simplify the code by keeping the values as plain bytes rather than
slices so that it’s explicit in their type that they are always just
one byte.

matklad committed:
test: simplify cops support in KeyValue runtime (#7213)

* eagarly compute assignemt of validators to blocks & chunks, which I find a bit easier to wrap my head around (you can print `EpochValidatorSet`s and just see the result)
* compress the code by centralizing cfg handling

akhi3030 committed:
Refactor: remove some dead code (#7216)

Closes #7203 

There are a few structs that do not need to derive `Serialize` and `Deserialize`.  Remove the derives highlights that there are some structs that we do not need at all and that we do not need to define custom deserializer: `bytes_as_string`.

Tests
-------
- `cargo check --tests --benches` does not complain.
- Other existing tests

jakmeier committed:
refactor: remove metric field in `GasCost` (#7215)

Explicitly tracking the metric inside `GasCost` is awkward for zero
values and it requires the estimator configuration to be passed around
more than strictly necessary.

This change makes the metric implicit, by virtue of having or not having
the measurement values stored inside.

For conversion to gas, and when printing, `ICount` measurements are
given priority over `Time` measurements when for both values exist.
This could be changed, if both need to be tracked simultaneously, for
now only one of them is of interest at the time.

posvyatokum committed:
export rocksdb cf size as prometheus metric (#6635)

Creating a function to get int rocks db properties for every column and adding a way to export it as a prometheus metric.
Reorganising parse_statistics to write stats into provided result param.

jakmeier committed:
refactor: `call_main` utility for integration test (#7217)

We have been copying the same code multiple times, this refactor
pulls out the functionality the `TestEnv`.

mina86 committed:
store: move RocskDB and TestDB structs to their separate modules (#7218)

This helps separate things which relates to the Database interface
from code that is RocksDB and TestDB specific.

Mic92 committed:
also respect storage config when checking rocksdb version (#7214)

In our case it's the number of file descriptors that is not respected.

jc-near committed:
Modify state dumper and replay to select / replace the right accounts (#7212)

This PR:
(1) makes sure that state dumper always include the treasury account in its output genesis file
(2) changes `pytest`-`replay` to make sure the genesis file it generates always passes the checks enforced by `neard`, since the list of staked accounts found in tries will be compared against the list validators.

### Test steps
 - Follow https://github.com/near/nearcore/blob/master/pytest/tests/replay/README.md to test it on mainnet

### Reviewers
 - @mzhangmzz 
 - @mm-near

jc-near committed:
show output file path after dumping tx (#7221)

This PR adds a line of output at the end of `dump_state`, for convenience. 

### Test steps
 - `make neard`
 - `/target/release/neard --home $NEAR_HOME_PATH view-state dump_tx --start-height $X --end-height $Y`
 - expect to see a line of output in the end that shows 

### Reviewers
 - @mzhangmzz 
 - @mm-near

nikurt committed:
Display block and chunk producers of an epoch (#7222)

* Display block and chunk producers of an epoch.

* Epoch length

* test: simplify mocking in setup_mock_all_validators (#7194)

There are many lovable aspects to `setup_mock_all_validators`! What this PR deals with is "tying the knot" infra. `setup_mock_validator` is based on creating `n` validators and setting up the mock network between them. To create a mock network, you need all `n` validators, and to create a validator, you need a mock network. The cycle is broken at runtime, in three different ways: 

* inside `setup_mock_all_validators`, we have `locked_connectors` 
* at some call-sites, we overwrite the whole `peer_manager_mock`
* yet at other call-sites, we essentially create a copy of `locked_connectors`

This PR unifies that to a single infra: 

* there's a single `Arc<OnceCell<Vec<Addr<Client>>>>` inside `setup_mock_all_validators`
* `OnceCell::wait` is used as a barrier 
* The `&[Addr<CLient>]` are passed in as an argument to `peer_manager_mock`, so that it doesn't have to re-invent a way to get a hold of those

This changes the signature of the function, so the diff is pretty large. It is designed to be reviewed per-commit.

* Fix m1 build issue (#7199)

* renamed Validator to AccountData (#7206)

https://nearinc.atlassian.net/browse/CPR-94

* feat: new estimator variance check (#7197)

The condition to check for high variance of measurements so far has been
that no single sample is further than 15% from the mean away. We often
missed that when running with 5 iterations and one warm-up iteration.
The counter-intuitive side of this check is that is becomes less likely
that variance is considered low enough, if more iterations are added.

The new condition computes actual variance and standard deviation.
Now we expect that standard deviation is lower than 10%.

Also add an outer loop to block overhead computation. Usually there
is a loop + a number of actions inside a block. But this estimation
has empty blocks and therefore each measurement so far was
just a single apply block call. Grouping 10 empty blocks together
for each measurement and adding an outer loop for `--iters` reduces 
the overall variance.

* test: more useful assert (#7211)

Got this one spuriously(?) firing in some test with nightly features.
Let's make the failure a bit more helpful here!

* trie_key: don’t store single bytes as slices (#7200)

trie_key module stores separators and keys as slices even though they
are all just one-byte long.  There are parts of the code which assume
the length to be one and won’t work if the length ever changes.

Simplify the code by keeping the values as plain bytes rather than
slices so that it’s explicit in their type that they are always just
one byte.

* test: simplify cops support in KeyValue runtime (#7213)

* eagarly compute assignemt of validators to blocks & chunks, which I find a bit easier to wrap my head around (you can print `EpochValidatorSet`s and just see the result)
* compress the code by centralizing cfg handling

* Refactor: remove some dead code (#7216)

Closes #7203 

There are a few structs that do not need to derive `Serialize` and `Deserialize`.  Remove the derives highlights that there are some structs that we do not need at all and that we do not need to define custom deserializer: `bytes_as_string`.

Tests
-------
- `cargo check --tests --benches` does not complain.
- Other existing tests

* refactor: remove metric field in `GasCost` (#7215)

Explicitly tracking the metric inside `GasCost` is awkward for zero
values and it requires the estimator configuration to be passed around
more than strictly necessary.

This change makes the metric implicit, by virtue of having or not having
the measurement values stored inside.

For conversion to gas, and when printing, `ICount` measurements are
given priority over `Time` measurements when for both values exist.
This could be changed, if both need to be tracked simultaneously, for
now only one of them is of interest at the time.

* export rocksdb cf size as prometheus metric (#6635)

Creating a function to get int rocks db properties for every column and adding a way to export it as a prometheus metric.
Reorganising parse_statistics to write stats into provided result param.

* refactor: `call_main` utility for integration test (#7217)

We have been copying the same code multiple times, this refactor
pulls out the functionality the `TestEnv`.

* Replace a map with a for-loop

Co-authored-by: Aleksey Kladov <aleksey.kladov@gmail.com>
Co-authored-by: Phuong Nguyen <ChaoticTempest@users.noreply.github.com>
Co-authored-by: pompon0 <pompon.pompon@gmail.com>
Co-authored-by: Jakob Meier <mail@jakobmeier.ch>
Co-authored-by: Michal Nazarewicz <mina86@mina86.com>
Co-authored-by: Akhilesh Singhania <akhi3030@gmail.com>
Co-authored-by: posvyatokum <posvyatokum@edu.hse.ru>

blasrodri committed:
address several comments

blasrodri committed:
rm reverse from get_proof

blasrodri committed:
Merge branch 'master' into blas-mpt-proof

blasrodri committed:
document get_proof

blasrodri committed:
return proof as type not as string

mina86 committed:
store: handle Insert operation in ShardTries::update_cache (#7234)

Due to the use of default case in a match, DBop::Insert case doesn’t
have explicit match branch in ShardTries::update_cache.  To be
consistent with Set and Delete handling, the code should be panicking
if inserting to ColState.

Remove the default case and replace conditional cases with `if`s
inside of the branches so that each operation need to be handled
explicitly.

mm-near committed:
Shardnet improvements to debug page (#7229)

- Added links to the network page
- Validators that are not reachable are marked as red.

blasrodri committed:
address comments

blasrodri committed:
docs improvement

blasrodri committed:
add deepsize feature

marcelo-gonzalez committed:
fix(sync): fix a syncing stall when a shard split is happening soon (#6939)

The client code will actually only update its sync state if
state_sync.run() returns Completed or Changed. If the sync state is
changed in StateSync::sync_shards_status() but the return value is
(false, false), then this won't happen. Note that forgetting to update
update_sync_status only really leads to problems when need_shard and
this_done are both left unset. But can't hurt to just set it to true
whenever we update the status in the code. Concretely, this bug leads
to a stall when syncing state if a shard split is incoming, with log
lines like the following showing up over and over:

022-05-29T17:16:12.216741Z DEBUG sync:run{sync="StateSync"}: sync: syncing state ..... status: StateDownloadComplete ...
...
...
022-05-29T17:16:12.216741Z DEBUG sync:run{sync="StateSync"}: sync: syncing state ..... status: StateDownloadComplete ...
...
...
022-05-29T17:16:12.216741Z DEBUG sync:run{sync="StateSync"}: sync: syncing state ..... status: StateDownloadComplete ...

This is sadly actually pretty hard to add a test for. A unit test
would be nice, but KeyValueRuntime doesn't update its shard layout, so
we can't actually get will_shard_layout_change_next_epoch() to return true
in the test code without adding that functionality. Also just using a
real NightshadeRuntime isn't possible since near-client doesn't import
nearcore, and it feels wrong to import that just for this test...

And then pytests are actually not easy to implement to get this to
fail either, since changing the shard layout in the genesis config and
setting the initial protocol version to 47 (like in
state_sync_fail.py) will cause the shard split to happen at epoch
height 2.  So to trigger this bug we want state sync to happen from
epoch height 1, but there's only actually a small corner case that
will allow that, where we start state sync when our header head is at
epoch height 2, but within 5 blocks of the epoch start. (this is actually
what happened when I ran into the stuck logs above)

jakmeier committed:
fix: graceful zero least-squares estimations (#7230)

Estimations such as `GasMeteringBase` can return a zero value, which has
no metric attached. The least-squares method should not panic in such
cases and instead treat the value as 0 for every metric.

regression introduced in #7215

blasrodri committed:
make deploy test contract optional

To remove non-determinism

mzhangmzz committed:
add metrics for block ordinal for current chain head (#7243)

This will be used to calculate number of skipped blocks. The current calculation of this metrics is problematic.

mina86 committed:
store: Trie: empty_root() → EMPTY_ROOT and use it more consistently (#7240)

* store: Trie: replace empty_root() with EMPTY_ROOT and use it more consistently

* StateRoot::default as well

jakmeier committed:
debug: forward debug flag to loading estimation (#7231)

--debug flag on estimator is used to show intermediate results and other
useful information. The contract loading estimation did not respect this
flag so far because the code predates the debug flag.

anannyenaik committed:
Fixed grammatical errors in `SECURITY.md` (#7250)

nikurt committed:
Expose node_key in the status response (#7239)

Context: https://nearinc.atlassian.net/browse/SHRD-115

nikurt committed:
fix(shardnet): Create the correct account (#7144)

* fix(shardnet): Create the correct account.

* fix(shardnet): Create the correct account.

* Update pytest/tests/shardnet/scripts/create_account.sh

Co-authored-by: Michal Nazarewicz <mina86@mina86.com>

* Update pytest/tests/shardnet/scripts/create_account.sh

Co-authored-by: Michal Nazarewicz <mina86@mina86.com>

* Update pytest/tests/shardnet/scripts/create_account.sh

Co-authored-by: Michal Nazarewicz <mina86@mina86.com>

* verbose output

Co-authored-by: Michal Nazarewicz <mina86@mina86.com>
Co-authored-by: near-bulldozer[bot] <73298989+near-bulldozer[bot]@users.noreply.github.com>

matklad committed:
Revert "feat: stabilize precharged contract loading gas (#7177)" (#7241)

This reverts commit f7aca03262bdbc58a8b1cedeb4deb82a5c565d01.

Reverting based on discussion at https://github.com/near/nearcore/issues/7227

The TL;DR is while the current impl of the feature charges correct cost for contract loading at *logically* correct place, we'd want to move the charging a little bit earlier physically, even before we load the contract from the database, so that the cost can be lower without adding an extra protocol feature here.

akhi3030 committed:
refactor: Use std API to avoid throwing errors that should not be possible (#7236)

`encode_into()` is using the `Cursor` API to write bytes to a vector.  This API can return errors which we have to handle.  Instead, the standard vector API does not return errors so we do not have handle errors in our code that are "impossible" to happen.

mina86 committed:
store: prefer Trie::iter to TrieIterator::new (#7237)

Co-authored-by: Aleksey Kladov <aleksey.kladov@gmail.com>

mina86 committed:
core: harmonise dec_format serialisation formats (#7248)

Replace all `{option_,}{u64,u128}_dec_format{,_compatible}` formats
with a single `dec_format` which automatically figures out the type
it’s serialising/deserialising.  This makes it somewhat easier to use
the formats since a simple `#[serde(with = "dec_format")]` now just
works without user having to think about the actual type of the field.

This brings one change in behaviour though.  Previously only some of
the fields used ‘compatible’ variant.  I.e. they would allow the field
to be either string or a number.  Most always expected a string.  With
this change, when deserialising u128 and u64 fields they can all be
gives as strings (the canonical format) or numbers.

Finally, remove one remaining unused instance of u128_dec_format which
was overlooked in previous commit.

Fixes: https://github.com/near/nearcore/issues/3384

nikurt committed:
Scraper for collecting IP addresses of validators (#7252)

https://nearinc.atlassian.net/browse/SHRD-115

pompon0 committed:
implemented accounts data cache (#7208)

https://nearinc.atlassian.net/browse/CPR-94

accounts_data::Cache is the central component collecting the SignedAccountData (data about validator nodes broadcasted by the the validators themselves). It takes care of preventing DDOS attacks of various types: minimizes the amount of signature verifications (which is computationally expensive) and detects malicious data as early as possible. It will be plugged into the PeerManager logic in the next PR.

matklad committed:
refactor: avoid double hash-map lookup (#7259)

mzhangmzz committed:
add more detailed metrics for block processing times (#7244)

* add more detailed metrics for block processing times

* address comments

Co-authored-by: near-bulldozer[bot] <73298989+near-bulldozer[bot]@users.noreply.github.com>

frol committed:
refactor: Make the near-client-primitives optional dependency in near-jsonrpc-primitives (#7242)

mina86 committed:
store: make TrieRefcountChange::rc non-zero (#7257)

bucanero committed:
Update README.md (#7262)

docs links

matklad committed:
refactor: simplify send message (#7258)

There are two reasons why we are doing this:

* The current setup with an associated function just doesn't make sense,
  this feels very much like a method.
* In the near future we want to split `connected_peers` into `tier1` and
  everyone else, it would help if sending a message didn't talk about
  connected_peers directly.

matklad committed:
refactor: remove unwrap (#7261)

pompon0 committed:
refactored NetworkRecipient interface (#7269)

Refactored NetworkRecipient interface so that it supports multiple message types and easy mocking. Currently we have an enum for all-messages and an enum for all-responses, which requires an unwrap of right response variant at every call site.

bucanero committed:
Update README.md (#7271)

docs links

mzhangmzz committed:
Add more detailed validator information in debug page (#7264)

<img width="1767" alt="Screen Shot 2022-07-25 at 8 14 52 PM" src="https://user-images.githubusercontent.com/34969888/180896324-08faee77-502a-450d-bcf7-50eca002b677.png">

bucanero committed:
Update README.md (#7274)

fix link

akhi3030 committed:
Refactor: remove unnecessary `pub` from a field (#7266)

Co-authored-by: near-bulldozer[bot] <73298989+near-bulldozer[bot]@users.noreply.github.com>

mina86 committed:
runtime: refactor try_fold to avoid unnecessary collect in check_balance (#7246)

Refactor .iter().map().collect().iter().try_fold() chains present in
check_balance function into a single invocation of .try_fold() with
a closure which incorporates all the necessary steps.  This not only
arguably simplifies the code but also removes .collect calls which
create unnecessary vectors.

akhi3030 committed:
refactor: replace some `allow(dead_code)` with `cfg(test)`s (#7268)

* refactor: replace some `allow(dead_code) with `cfg(test)`s

* apply PR feedback

Co-authored-by: near-bulldozer[bot] <73298989+near-bulldozer[bot]@users.noreply.github.com>

mina86 committed:
runtime: Arc<CompiledContractCache> → Box<_> (#7276)

The Arc<CompiledContractCache> are never cloned.  Simplify code by
just using a Box instead.

nikurt committed:
Extend TelemetryInfo with an additional freeform field (#7270)

The new text field is ignored by the explorer.
Expose block production delays in the freeform field.

https://nearinc.atlassian.net/browse/SHRD-116

mina86 committed:
store: use Database directly in StoreCompiledContractCache (#7275)

The main motivation here isn’t obvious.  It’s to decouple the cache
from Store struct similarly how peer store is now talking to Database
directly.  Long-term plan is to have hot and cold storage;  it’s
beneficial if part of the code which only ever talk to hot storage
don’t need to go through Store interface.

As a side benefit, this reduces the number of clones when storing
contracts in the cache.

marcelo-gonzalez committed:
refactor: don't compute the genesis json hash if it's already stored (#7273)

This computation takes several minutes if there are many records, like
in testnet, or a network with state forked from mainnet. And then if
the hash was already stored, the result is promptly discarded.  If the
TODO here around re-enabling the hash check gets done then we will
have to revisit this.

mina86 committed:
Update to rustc 1.62.1 (#7284)

I’ve been getting ICEs when testing some changes on NayDuck.  Updating
to 1.62.1 fixes the issue.  I didn’t look into it further.  Let’s just
update the compiler.

mzhangmzz committed:
Fix some bug in validator info debug page (#7283)

* fix validator info

* fix

* add next epoch validators

Co-authored-by: near-bulldozer[bot] <73298989+near-bulldozer[bot]@users.noreply.github.com>

nikurt committed:
fix(collect-ips): Use non-debug endpoints (#7279)

In shardnet still can't reach all validators:
```
Validators found: 122.
Visited nodes: 917.
```

mm-near committed:
[SHRD-134] Exporting more detailed info about peers (#7289)

* Export more information about peers (network stats etc)

* fmt

* fix field name

Co-authored-by: near-bulldozer[bot] <73298989+near-bulldozer[bot]@users.noreply.github.com>

mina86 committed:
network: get rid of RoutedMessageBody::Query{Request,Response} (#7282)

We haven’t been sending those messages for a while now so support for
borsh serialising and deserialising them is unnecessary.  Meanwhile,
the serialisation forces bunch of other types to be borsh-serialisable
as well making it harder to change those types.  And changing them is
something we might want to do.

The message used to be sent when view client couldn’t handle a Query
due to lack of state.  In that case it would route the message to
other nodes.  This feature has been removed in commit f68140e65c48:
‘Add structured errors to query method in ViewClient and RPC.’

matklad committed:
doc: document why we may need to store more than one outcome per tx (#7292)

matklad committed:
test: introduce ValidatorSchedule to specify COPs for KeyValue runtime conveniently (#7256)

With cops, we now need to somehow configure `setup_mock_all_validators` to also specify cops. It *could* be done by adding just another argument, but I'd rather avoid that. Instead, what I suggest is bunching up all related configs into a single builder object, `ValidatorSchedule`, and use that as a universal tool to configure validators for KeyValueRuntime. 

Note that this PR doesn't change `setup_mock_all_validators` -- that's going to be a bigger diff, so I want to land that separately.

I might also merge this with `EpochValidatorSet`, the structs handle similar data, but, at this stage, `EpochValidatorSet` is implementation detail, while `ValidatorSchedule` is the API for authors of tests. 

Can be reviewed per commit.

mm-near committed:
[SHRD-118] Added more information to the network_info debug page. (#7286)

* info about current chunk producers (and how many are reachable)
* list of unreachable chunk/block producers
* added colors to height
* added information about the type of connection (IN vs OUT)
* added columns about first connection time and last ping (they will start working once the changes to the neard are submitted).

<img width="1398" alt="image" src="https://user-images.githubusercontent.com/91919554/181454724-62b3bb7b-682b-4f1a-93a0-9e1d35f57bd6.png">

pompon0 committed:
Implemented broadcasting per-account data (#7171)

Specifically this PR implements broadcasting data by validator accounts for the current epoch. Currently it just includes the list of IPs under which the validator is reachable. It will be used to establish TIER1 network. This is just implementation of the broadcast. Seeding the data at the validators will be implemented in the next PR.

jira issue: https://nearinc.atlassian.net/browse/CPR-94

blasrodri committed:
add helper func to avoid dealing with large proofs manually

mina86 committed:
store: deduplicate code retrieving of a raw trie node (#7233)

Introduce Trie::retrieve_raw_node method and use it in three places
which had practically identical code.

pompon0 committed:
added signing of AccountData to ValidatorSigner (#7272)

and adjusted the accounts_data::Cache accordingly.

akhi3030 committed:
minor improvements to CONTRIBUTING.md (#7294)

matklad committed:
test: use ValidatorSchedule in setup_mock_all_validators (#7280)

This'll allow configuring chunk-only producers for tests using this infra

follow up to #7256

matklad committed:
docs: add links to NEP and implementation PR to feature-stabilization checklist (#7288)

mina86 committed:
jsonrpc: fuzz: fix Base58String type; should read 64 (#7302)

While at it, convert the fuzz target to use encoders from near
primitives rather than using base64 and bs58 crates directly.

vmarkushin committed:
`verify_proof` for non-membership of a value (WIP)

blasrodri committed:
remove TrieProofItem in favor of RawTrieNode

add non membership proof test
encode nodes as string (base64)

mina86 committed:
nearcore: deprecate network configurations’s external_address field (#7300)

The field has never been used.  It’s only causes confusion as to its
purpose.  Deprecate it.

Issue: https://github.com/near/nearcore/issues/7198

blasrodri committed:
avoid unwraping on serialization of the proof

blasrodri committed:
enhance docs of get_proof

blasrodri committed:
more changes (wip)

blasrodri committed:
use RawTrieNodeWithSize instead

blasrodri committed:
rm unused deps

blasrodri committed:
serialize not needed

blasrodri committed:
correct docs for ViewStateResult

blasrodri committed:
simplify code + add non membership tests

mina86 committed:
neard: use release tags when determining build string (#7305)

Because 1.x.y tags we’re using are not annotated, git describe ignores
them when describing a commit.  Add --tags flag to make it take a look
at such tags.  While at it, include --match=[0-9]* so that the command
excludes non-nearcore-release tags such as crates-0.14.0.

mina86 committed:
Prefer base64 from near_primitives; this puts everyone on same version (#7301)

enter the commit message for your changes. Lines starting

blasrodri committed:
Update core/store/src/trie/mod.rs

Co-authored-by: Vladislav <negigic@gmail.com>
Signed-off-by: Blas Rodriguez Irizar <rodrigblas@gmail.com>

blasrodri committed:
Merge pull request #9 from ComposableFi/vmarkushin/trie-nm-proof

Vmarkushin/trie nm proof

blasrodri committed:
change not_found to found

mzhangmzz committed:
fix a bug in send_network_chain_info (#7309)

The get_validator_info function doesn't work with the current epoch id, see comments in ValidatorIdentifier.

mina86 committed:
pytest: update tests to no longer used deleted ‘testnet’ command (#7311)

Issue: https://github.com/near/nearcore/issues/4372

mina86 committed:
core: views: prefer base64_format to manual base64 conversions (#7308)

Rather than declaring a field String and having all the users do
base64 conversion manually, declare it Vec<u8> with base64_format
serialisation scheme.  This way, only the serialisation code needs to
worry how the data looks on the wire.  This also means that users
reading the data don’t need to worry about decoding errors.

Fixes: https://github.com/near/nearcore/issues/7232

mina86 committed:
doc: prefer for loops to for_each and try_for_each methods (#7255)

* doc: prefer for loops to for_each and try_for_each methods

* Update chain/chunks/src/lib.rs

Co-authored-by: Akhilesh Singhania <akhi@near.org>

* Update runtime/runtime-params-estimator/estimator-warehouse/src/check.rs

Co-authored-by: Akhilesh Singhania <akhi@near.org>

* fix

* a few more

* some better wording

Co-authored-by: Akhilesh Singhania <akhi@near.org>

blasrodri committed:
address comments

blasrodri committed:
Merge remote-tracking branch 'near/master' into blas-mpt-proof

blasrodri committed:
adjust master changes

blasrodri committed:
address several comments from Akhi

blasrodri committed:
introduce ProofState

blasrodri committed:
include comments

blasrodri committed:
comments on interface

blasrodri committed:
make verify proof a standalone function

blasrodri committed:
add comments

blasrodri committed:
validate size

blasrodri committed:
simplify comparison

blasrodri committed:
more comments

blasrodri committed:
simplify get_proof
